library/core/registry/src/main/java/org/quiltmc/qsl/registry/api/dynamic/DynamicMetaregistry.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2023 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.registry.api.dynamic;
+
+import com.mojang.serialization.Codec;
+import net.minecraft.registry.Registry;
+import net.minecraft.registry.RegistryKey;
+import net.minecraft.util.Identifier;
+import org.quiltmc.qsl.registry.impl.dynamic.DynamicMetaregistryImpl;
+
+/**
+ * Allows registration of dynamic registries for use through {@link net.minecraft.registry.DynamicRegistryManager}.
+ *
+ * <p>Dynamic registries are only available in a world context. Entries must be added either:<ul>
+ * <li>as datapack files
+ * <li>through the {@link org.quiltmc.qsl.registry.api.event.RegistryEvents#DYNAMIC_REGISTRY_SETUP dynamic registry setup event}
+ * </ul>
+ *
+ * <p>This registry will be frozen at the same time as static registries.
+ */
+public final class DynamicMetaregistry {
+	/**
+	 * Registers a serverside dynamic registry.
+	 *
+	 * <p>Entries will be loaded from {@code "data/<namespace>/<registry_path>"} for every datapack {@code namespace},
+	 * where {@code registry_path}'s value is {@code key.getLocation().getPath()}.
+	 * The registry's own namespace is essentially ignored when loading values,
+	 * meaning <strong>the registry {@code key}'s path must be unique by itself</strong>
+	 * (e.g. {@code RegistryKey.ofRegistry(new Identifier("<mod_id>", "<mod_id>_<resource_name>))}).
+	 *
+	 * @param <E>        the type of elements in the dynamic registry
+	 * @param key        a {@link RegistryKey#ofRegistry(Identifier) key for the new dynamic registry}
+	 * @param entryCodec the codec used to deserialize entries from datapacks
+	 * @throws IllegalStateException if this registry of registries already got frozen
+	 */
+	public static <E> void register(RegistryKey<? extends Registry<E>> key, Codec<E> entryCodec) {
+		DynamicMetaregistryImpl.register(key, entryCodec);
+	}
+
+	/**
+	 * Registers a dynamic registry which contents get synced between the server and connected clients.
+	 *
+	 * <p>Entries will be loaded from {@code "data/<namespace>/<registry_path>"} for every datapack {@code namespace},
+	 * where {@code registry_path}'s value is {@code key.getLocation().getPath()}.
+	 * The registry's own namespace is essentially ignored when loading values,
+	 * meaning <strong>the registry {@code key}'s path must be unique by itself</strong>
```

Comment:
oh no, I think that's one of the reasons why the original PR stalled; we actually wanted a mechanism where if the namespace is omitted, it's assumed to be `minecraft`

Reply:
Maybe the javadoc isn't clear, or maybe I didn't understand you - what do you want to default to `minecraft` exactly ? The issue I'm trying to allude to is that `mod1:dynamic_registry` and `mod2:dynamic_registry` are loaded with the path `data/<every_namespace_that_exists>/dynamic_registry`

Reply:
ah, what I mean is that for modded dynamic registries, you should specify a namespace, or else `data/<namespace>/dynamic_registry` is assumed to refer to `minecraft:dynamic_registry` and therefore it would be equivalent to `data/<namespace>/minecraft/dynamic_registry`;
There has been convos before about having this implemented in general (mainly due to how REA has explicit namespaces but it doesn't allow to omit `minecraft` for vanilla registries), and yeah, this was a conclusion that we had, and yeah, I believe that if we end avoiding problems where registries must be like `mod_id:mod_id_gases`? it would be great

Reply:
So if I understand correctly, the solution you suggest would be to systematically prepend the registry path with `<registry_namespace>/`, and to patch registry loading such that vanilla registries can be loaded from `minecraft/<registry_path>` ?

Reply:
This seems like a good system; it's also very similar to the system that my PR (#264) uses, so consistency is always nice.

Reply:
> So if I understand correctly, the solution you suggest would be to systematically prepend the registry path with `<registry_namespace>/`, and to patch registry loading such that vanilla registries can be loaded from `minecraft/<registry_path>` ?

I needed a discussion on Discord in order to clear my mind more; Conclusion? Yep, this would bring the functionality that I've been thinking of

Reply:
Looking again at the code, the path from which resources are loaded is tightly coupled to the key. I therefore see 3 possibilities for implementing this:
1. Patch the registry loading code such that resources for the key `<namespace>:<path>` are always loaded from `<namespace>/<path>`, with a hardcoded exception for the Minecraft namespace. Pro: mods can just use `RegistryKey.ofRegistry(new Identifier("<namespace>", "<path>"))`. Con: immediately incompatible with every mod that already registers dynamic registries (e.g. [Blabber](https://github.com/Ladysnake/Blabber)).
2. Force the registry keys used to register new dynamic registries to have a `<namespace>:<namespace>/<path>` structure. Pros: least effort, transparent for mods. Con: not statically verified.
3. Make all registration methods take an `Identifier` and return a `RegistryKey` built as `<namespace>:<namespace>/<path>`. Pro: statically ensures the path is correct. Con: those keys are usually stored as constants, meaning the registration would have to be done during static initialization and would become part of a mod's public API.

By myself I would lean towards option 2 as it seems the least intrusive. What do you think ?

Reply:
Hm, I'd be heavily inclined to Option #1 due to it looking like the best solution, with the mod breaking not being a concern as long as the ecosystem could move to it; Problem: unfortunately, we share the ecosystem with this hecker: *looks at fabric*;

Unfortunately, we'll have to compromise then with a solution that is backwards-compatible, which yeah, is complicated; I don't like the idea of having RegistryKeys not be the straightforward `<namespace>:<path>`, so I'm discarding options #2 and #3;

I do wonder if we could adopt option #1 and like, have some sort of solution on QFAPI's side that attempts to add exceptions for Fabric mods, similar to how QSL has content registries based on REA but QFAPI has deferring queues in order to adjust messy attempts to add to them by Fabric mods; I am having a hard time to think about how this exception system could be done though, since it could end effectively destroying the benefits of this system if done wrong; if done right, however, this option #1.5 would be the best option here, I believe

Reply:
Fair enough, I went ahead and implemented what I think is a decent spin on option 1 - making it so only dynamic registries registered through our API get affected by the path changes.